service/controller/resource/tccpn/create.go

```diff
@@ -0,0 +1,515 @@
+
+		} else if *o.Stacks[0].StackStatus == cloudformation.StackStatusCreateInProgress {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the tenant cluster's control plane nodes cloud formation stack has stack status %#q", cloudformation.StackStatusCreateInProgress))
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
+		} else if *o.Stacks[0].StackStatus == cloudformation.StackStatusUpdateInProgress {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("the tenant cluster's control plane nodes cloud formation stack has stack status %#q", cloudformation.StackStatusUpdateInProgress))
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return nil
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "found the tenant cluster's control plane nodes cloud formation stack already exists")
+	}
+
+	{
+		update, err := r.detection.ShouldUpdate(ctx, cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if update {
+			err = r.updateStack(ctx, cr)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) createStack(ctx context.Context, cr infrastructurev1alpha2.AWSControlPlane) error {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var templateBody string
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "computing the template of the tenant cluster's control plane nodes cloud formation stack")
+
+		params, err := newTemplateParams(ctx, cr, r.encrypterBackend, r.apiWhitelist)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		templateBody, err = template.Render(params)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "computed the template of the tenant cluster's control plane nodes cloud formation stack")
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "requesting the creation of the tenant cluster's control plane nodes cloud formation stack")
+
+		i := &cloudformation.CreateStackInput{
+			Capabilities: []*string{
+				aws.String(capabilityNamesIAM),
+			},
+			EnableTerminationProtection: aws.Bool(true),
+			StackName:                   aws.String(key.StackNameTCCPN(&cr)),
+			Tags:                        r.getCloudFormationTags(cr),
+			TemplateBody:                aws.String(templateBody),
+		}
+
+		_, err = cc.Client.TenantCluster.AWS.CloudFormation.CreateStack(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "requested the creation of the tenant cluster's control plane nodes cloud formation stack")
+	}
+
+	return nil
+}
+
+func (r *Resource) getCloudFormationTags(cr infrastructurev1alpha2.AWSControlPlane) []*cloudformation.Tag {
+	tags := key.AWSTags(&cr, r.installationName)
+	tags[key.TagStack] = key.StackTCCPN
```

Comment:
I think I know again why I added the control plane tag in the first place. I wanted to treat it as Node Pool. We could even add `ControlPlane` CRs to a Cluster and run multiple Node Pool like stacks for the Control Plane Nodes. That way we have a mechanism to more easily refactor the underlying infrastructure in case we have to. This would go towards the efforts of preventing breaking changes. We would just need to add unique IDs to the Control Plane CRs. /cc @giantswarm/sig-operator

Reply:
I like where this is going, however, as we are tying etcd into the CP machines an actual switch from one to another CP might be more complex, with separate etcd machines that might be easier.

Or am I mistaking your "mechanism to more easily refactor the underlying infrastructure" for what it is? Maybe you mean the CP stacks hand over control of the underlying infra without actually creating new infra?

Reply:
Treating the new TCCPN (Tenant Cluster Control Plane Nodes) stack like a Node Pool means in theory we can add and remove them easily, because the underlying infrastructure becomes more pluggable. When you have one lego block plugged in you are usually able to replicate that. You are right there is additional complexity to be aware of in terms of etcd but I would like to aim for a design where you create the G8sControlPlane/AWSControlPlane CR "combo" to manage one set of master nodes and when you create another "combo" you add another 1 or 3 masters to the Tenant Cluster. Then you could tear down the old one and have effectively replaced that piece of infrastructure the old stack was running. I only want to make sure we thought about this and have the "combo" identified uniquely. Simply to be able to replicate it later on in case we want/have to.